### ISSUE TITLE
WARNING Command ['start', 'dovecot'] has failed. Received RuntimeError("can't start new thread",) 

### ISSUE THREAD

**Issue by AndrzejL-eu** [fail2ban/fail2ban#1336]
Hi,

[root@andrzejl fail2ban-git]# grep -i warning /var/log/fail2ban.log
2016-02-24 18:35:09,774 fail2ban.transmitter [1182]: WARNING Command ['stop'] has failed. Received RuntimeError('cannot join thread before it is started',)
2016-02-24 18:35:12,088 fail2ban.transmitter [11393]: WARNING Command ['start', 'postfix-sasl'] has failed. Received RuntimeError("can't start new thread",)
2016-02-24 18:35:12,092 fail2ban.transmitter [11393]: WARNING Command ['start', 'dovecot'] has failed. Received RuntimeError("can't start new thread",)
[root@andrzejl fail2ban-git]# pacman -Q | grep systemd
libsystemd 229-3
python-systemd 231-1
systemd 229-3
systemd-sysvcompat 229-3
[root@andrzejl fail2ban-git]# pacman -Q | grep fail2ban
fail2ban-git 0.9.3.r180.gbd822d0-1
[root@andrzejl fail2ban-git]#

Arch Linux 32 bit, fail2ban fails to start properly possibly after systemd upgrade. Hitting RuntimeError...

Cheers.

Andrzej

P.S. Being overtired I've first started this bug report at SF several days ago. Only after a proper R&R I've noticed the mistake... That's what I get for trying to pull extra shifts ;)...

**Comment by kakra** [comment]
This looks like #969...

**Comment by AndrzejL-eu** [comment]
Yes it does but I was trying not to necro-post. Last night - I recompiled git master and now it happens (the warning) with 0 jails opened. None of my jails opened. All hit the runtime error.

Cheers.

Andrzej

**Comment by sebres** [comment]
I think, I've already seen such errors in early versions of python on some platforms. Back then the reason for this was for the most part a little bit buggy (threaded) implementation of stdout- and stderr-pipes, by executing of the process via `subprocess.Popen` in python.
Which python version?

Can also provide more information. Use `loglevel = DEBUG` in fail2ban.local and start f2b. 
Dump of jail parameters with `fail2ban-client -d | grep dovecot`...

Should it be really `executeCmd` by start/stop of jail, then you can try additionally my following branch (prev. #1321 etc.), where I've fully rewritten the starting of sub processes using executeCmd.
https://github.com/sebres/fail2ban/tree/f2b-perfom-prepare-716-cs

**Comment by AndrzejL-eu** [comment]
[root@andrzejl andrzejl]# pacman -Q | grep python
python 3.5.1-1
python-dbus 1.2.0-5
python-dbus-common 1.2.0-5
python-gobject 3.18.2-1   
python-lxml 3.5.0-1
python-packaging 16.5-1   
python-pyinotify 0.9.6-2  
python-pyparsing 2.1.0-1  
python-pyqt4 4.11.4-4
python-setuptools 1:20.2.2-1
python-sip 4.16.9-2
python-six 1.10.0-1
python-sqlalchemy 1.0.12-1
python-systemd 231-1
python2 2.7.11-1
python2-beaker 1.8.0-1
python2-cairo 1.10.0-2
python2-cffi 1.5.2-1
python2-crypto 2.6.1-3
python2-cryptography 1.2.3-1
python2-dbus 1.2.0-5
python2-enum34 1.1.1-1
python2-feedparser 5.2.1-1
python2-gobject2 2.28.6-12
python2-idna 2.0-2
python2-ipaddress 1.0.16-1
python2-mako 1.0.3-1
python2-markupsafe 0.23-3
python2-notify 0.1.1-13
python2-packaging 16.5-1
python2-pillow 3.1.1-1
python2-pip 8.0.3-1
python2-ply 3.8-1
python2-pyasn1 0.1.9-1
python2-pycparser 2.14-2
python2-pyinotify 0.9.6-2
python2-pyopenssl 0.15.1-2
python2-pyparsing 2.1.0-1
python2-pysqlite 2.8.1-2
python2-pysqlite-legacy 1.0.1-3
python2-reportlab 3.3.0-1
python2-setuptools 1:20.2.2-1
python2-six 1.10.0-1
python2-twisted 15.5.0-2
python2-zope-interface 4.1.3-2
[root@andrzejl andrzejl]#

[root@andrzejl andrzejl]# updatedb
[root@andrzejl andrzejl]# locate fail2ban.local
[root@andrzejl andrzejl]# 

[root@andrzejl andrzejl]# grep loglevel /etc/fail2ban/fail2ban.conf 
# loglevel = DEBUG
# Option: loglevel
loglevel = DEBUG
[root@andrzejl andrzejl]# 

http://pastebin.com/raw/C1S3De2L

[root@andrzejl andrzejl]# fail2ban-client -d | grep dovecot
['add', 'dovecot', 'systemd']
['set', 'dovecot', 'findtime', 900]
['set', 'dovecot', 'addignoreip', '127.0.0.1/8']
['set', 'dovecot', 'addignoreip', '192.168.0.0/24']
['set', 'dovecot', 'addignoreip', '192.168.1.0/24']
['set', 'dovecot', 'addignoreip', '86.44.169.199']
['set', 'dovecot', 'addignoreip', '83.71.47.150']
['set', 'dovecot', 'addignoreip', '80.229.149.31']
['set', 'dovecot', 'logencoding', 'auto']
['set', 'dovecot', 'maxretry', 1]
['set', 'dovecot', 'usedns', 'warn']
['set', 'dovecot', 'ignorecommand', '']
['set', 'dovecot', 'bantime', 86400]
['set', 'dovecot', 'addfailregex', '^\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?(auth|dovecot(-auth)?|auth-worker)(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?(auth|dovecot(-auth)?|auth-worker)(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*(pam_unix(\\(dovecot:auth\\))?:)?\\s+authentication failure; logname=\\S* uid=\\S* euid=\\S* tty=dovecot ruser=\\S* rhost=<HOST>(\\s+user=\\S*)?\\s*$']
['set', 'dovecot', 'addfailregex', '^\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?(auth|dovecot(-auth)?|auth-worker)(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?(auth|dovecot(-auth)?|auth-worker)(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*(pop3|imap)-login: (Info: )?(Aborted login|Disconnected)(: Inactivity)? \\(((auth failed, \\d+ attempts)( in \\d+ secs)?|tried to use (disabled|disallowed) \\S+ auth)\\):( user=<\\S*>,)?( method=\\S+,)? rip=<HOST>(, lip=(\\d{1,3}\\.){3}\\d{1,3})?(, TLS( handshaking(: SSL_accept\\(\\) failed: error:[\\dA-F]+:SSL routines:[TLS\\d]+_GET_CLIENT_HELLO:unknown protocol)?)?(: Disconnected)?)?(, session=<\\S+>)?\\s*$']
['set', 'dovecot', 'addfailregex', '^\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?(auth|dovecot(-auth)?|auth-worker)(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?(auth|dovecot(-auth)?|auth-worker)(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*(Info|dovecot: auth\\(default\\)|auth-worker\\(\\d+\\)): pam\\(\\S+,<HOST>\\): pam_authenticate\\(\\) failed: (User not known to the underlying authentication module: \\d+ Time\\(s\\)|Authentication failure \\(password mismatch\\?\\))\\s*$']
['set', 'dovecot', 'addfailregex', '^\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?(auth|dovecot(-auth)?|auth-worker)(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?(auth|dovecot(-auth)?|auth-worker)(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*(auth|auth-worker\\(\\d+\\)): (pam|passwd-file)\\(\\S+,<HOST>\\): unknown user\\s*$']
['set', 'dovecot', 'addjournalmatch', '_SYSTEMD_UNIT=dovecot.service']
['set', 'dovecot', 'addaction', 'shorewall']
['set', 'dovecot', 'action', 'shorewall', 'actioncheck', '']
['set', 'dovecot', 'action', 'shorewall', 'actionunban', 'shorewall allow <ip>']
['set', 'dovecot', 'action', 'shorewall', 'actionstop', '']
['set', 'dovecot', 'action', 'shorewall', 'actionban', 'shorewall <blocktype> <ip>']
['set', 'dovecot', 'action', 'shorewall', 'actionstart', '']
['set', 'dovecot', 'action', 'shorewall', 'blocktype', 'reject']
['set', 'dovecot', 'action', 'shorewall', 'known/blocktype', 'reject']
['set', 'dovecot', 'addaction', 'sendmail-whois']
['set', 'dovecot', 'action', 'sendmail-whois', 'actioncheck', '']
['set', 'dovecot', 'action', 'sendmail-whois', 'actionunban', '']
['set', 'dovecot', 'action', 'sendmail-whois', 'actionstop', 'printf %b "Subject: [Fail2Ban] <name>: stopped on `uname -n`\nDate: `LC_ALL=C date +"%a, %d %h %Y %T %z"`\nFrom: <sendername> <<sender>>\nTo: <dest>\\n\nHi,\\n\nThe jail <name> has been stopped.\\n\nRegards,\\n\nFail2Ban" | /usr/bin/sendmail -f <sender> <dest>']
['set', 'dovecot', 'action', 'sendmail-whois', 'actionban', 'printf %b "Subject: [Fail2Ban] <name>: banned <ip> from `uname -n`\nDate: `LC_ALL=C date +"%a, %d %h %Y %T %z"`\nFrom: <sendername> <<sender>>\nTo: <dest>\\n\nHi,\\n\nThe IP <ip> has just been banned by Fail2Ban after\n<failures> attempts against <name>.\\n\\n\nHere is more information about <ip> :\\n\n`/usr/bin/whois <ip> || echo missing whois program`\\n\nRegards,\\n\nFail2Ban" | /usr/bin/sendmail -f <sender> <dest>']
['set', 'dovecot', 'action', 'sendmail-whois', 'actionstart', 'printf %b "Subject: [Fail2Ban] <name>: started on `uname -n`\nDate: `LC_ALL=C date +"%a, %d %h %Y %T %z"`\nFrom: <sendername> <<sender>>\nTo: <dest>\\n\nHi,\\n\nThe jail <name> has been started successfully.\\n\nRegards,\\n\nFail2Ban" | /usr/bin/sendmail -f <sender> <dest>']
['set', 'dovecot', 'action', 'sendmail-whois', 'known/sender', 'fail2ban']
['set', 'dovecot', 'action', 'sendmail-whois', 'known/known/dest', 'root']
['set', 'dovecot', 'action', 'sendmail-whois', 'known/sendername', 'Fail2Ban']
['set', 'dovecot', 'action', 'sendmail-whois', 'dest', 'andrzejl@icsserver.loc']
['set', 'dovecot', 'action', 'sendmail-whois', 'sendername', 'Fail2Ban']
['set', 'dovecot', 'action', 'sendmail-whois', 'known/known/sendername', 'Fail2Ban']
['set', 'dovecot', 'action', 'sendmail-whois', 'known/known/sender', 'fail2ban']
['set', 'dovecot', 'action', 'sendmail-whois', 'known/dest', 'root']
['set', 'dovecot', 'action', 'sendmail-whois', 'name', 'Dovecot']
['set', 'dovecot', 'action', 'sendmail-whois', 'known/name', 'default']
['set', 'dovecot', 'action', 'sendmail-whois', 'sender', 'fail2ban@icsserver.loc']
[root@andrzejl andrzejl]# 

Cheers,

Andrzej

**Comment by sebres** [comment]
3.5.1 o_O
I must come back later ... (too busy currently)

**Comment by kakra** [comment]
In case this helps debugging: I've run it on a full 32-bit system and could only at most add one jail... Add another and it fails with the thread error. Now I've booted the same server with a 64-bit kernel (still pure 32-bit userland): I can now add at most two jails before the error happens - additionally `/usr/bin/python3.4` sometimes segfaults:

```
# coredumpctl dump 22628
           PID: 22628 (fail2ban-server)
           UID: 0 (root)
           GID: 0 (root)
        Signal: 6 (ABRT)
     Timestamp: Sa 2016-03-05 13:07:54 CET (23min ago)
  Command Line: /usr/bin/python3.4 /usr/bin/fail2ban-server -s /run/fail2ban/fail2ban.sock -p /run/fail2ban/fail2ban.pid -x -b
    Executable: /usr/bin/python3.4
 Control Group: /system.slice/fail2ban.service
          Unit: fail2ban.service
         Slice: system.slice
       Boot ID: 3fed91a509764f029ef8dbe62e5aa5ac
    Machine ID: 3220c025abd7f15126417ecc56c92790
      Hostname: weser.netactive.de
       Message: Process 22628 (fail2ban-server) of user 0 dumped core.
                
                Stack trace of thread 22636:
                #0  0x00000000e703fa20 __kernel_vsyscall (linux-gate.so.1)
                #1  0x00000000e6c70469 __GI_raise (libc.so.6)
                #2  0x00000000e6c71bdd __GI_abort (libc.so.6)
                #3  0x00000000e6c6903a __assert_fail_base (libc.so.6)
                #4  0x00000000e6c690c7 __GI___assert_fail (libc.so.6)
                #5  0x00000000e44533a6 _gcry_ath_mutex_lock (libgcrypt.so.11)
                #6  0x00000000e4451074 _gcry_secmem_malloc (libgcrypt.so.11)
                #7  0x00000000e44505c9 _gcry_private_malloc_secure (libgcrypt.so.11)
                #8  0x00000000e444c86a do_malloc (libgcrypt.so.11)
                #9  0x00000000e444d2a5 _gcry_malloc_secure (libgcrypt.so.11)
                #10 0x00000000e4460b58 md_open (libgcrypt.so.11)
                #11 0x00000000e4461270 _gcry_md_open (libgcrypt.so.11)
                #12 0x00000000e4449f86 gcry_md_open (libgcrypt.so.11)
                #13 0x00000000e454d4ce journal_file_hmac_setup (libsystemd.so.0)
                #14 0x00000000e4548c32 journal_file_open (libsystemd.so.0)
                #15 0x00000000e453f994 add_any_file (libsystemd.so.0)
                #16 0x00000000e453fbbb add_file (libsystemd.so.0)
                #17 0x00000000e4540dd2 process_inotify_event (libsystemd.so.0)
                #18 0x00000000e4543d32 sd_journal_wait (libsystemd.so.0)
                #19 0x00000000e45816a6 Reader_wait (_reader.cpython-34.so)
                #20 0x00000000e6e7e156 PyCFunction_Call (libpython3.4.so.1.0)
                #21 0x00000000e6ee3205 call_function (libpython3.4.so.1.0)
                #22 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #23 0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #24 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #25 0x00000000e6e67267 function_call (libpython3.4.so.1.0)
                #26 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #27 0x00000000e6ee339d ext_do_call (libpython3.4.so.1.0)
                #28 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #29 0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #30 0x00000000e6ee1e8b fast_function (libpython3.4.so.1.0)
                #31 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #32 0x00000000e6e67267 function_call (libpython3.4.so.1.0)
                #33 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #34 0x00000000e6e526b2 method_call (libpython3.4.so.1.0)
                #35 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #36 0x00000000e6edcbdc PyEval_CallObjectWithKeywords (libpython3.4.so.1.0)
                #37 0x00000000e6f1701b t_bootstrap (libpython3.4.so.1.0)
                #38 0x00000000e6df52b9 start_thread (libpthread.so.0)
                #39 0x00000000e6d3621e __clone (libc.so.6)
                
                Stack trace of thread 22637:
                #0  0x00000000e703fa20 __kernel_vsyscall (linux-gate.so.1)
                #1  0x00000000e6d2dd91 __select (libc.so.6)
                #2  0x00000000e685db8c floatsleep (time.cpython-34.so)
                #3  0x00000000e6e7e156 PyCFunction_Call (libpython3.4.so.1.0)
                #4  0x00000000e6ee3205 call_function (libpython3.4.so.1.0)
                #5  0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #6  0x00000000e6e67267 function_call (libpython3.4.so.1.0)
                #7  0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #8  0x00000000e6ee339d ext_do_call (libpython3.4.so.1.0)
                #9  0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #10 0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #11 0x00000000e6ee1e8b fast_function (libpython3.4.so.1.0)
                #12 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #13 0x00000000e6e67267 function_call (libpython3.4.so.1.0)
                #14 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #15 0x00000000e6e526b2 method_call (libpython3.4.so.1.0)
                #16 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #17 0x00000000e6edcbdc PyEval_CallObjectWithKeywords (libpython3.4.so.1.0)
                #18 0x00000000e6f1701b t_bootstrap (libpython3.4.so.1.0)
                #19 0x00000000e6df52b9 start_thread (libpthread.so.0)
                #20 0x00000000e6d3621e __clone (libc.so.6)
                
                Stack trace of thread 22634:
                #0  0x00000000e703fa20 __kernel_vsyscall (linux-gate.so.1)
                #1  0x00000000e6d2dd91 __select (libc.so.6)
                #2  0x00000000e685db8c floatsleep (time.cpython-34.so)
                #3  0x00000000e6e7e156 PyCFunction_Call (libpython3.4.so.1.0)
                #4  0x00000000e6ee3205 call_function (libpython3.4.so.1.0)
                #5  0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #6  0x00000000e6e67267 function_call (libpython3.4.so.1.0)
                #7  0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #8  0x00000000e6ee339d ext_do_call (libpython3.4.so.1.0)
                #9  0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #10 0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #11 0x00000000e6ee1e8b fast_function (libpython3.4.so.1.0)
                #12 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #13 0x00000000e6e67267 function_call (libpython3.4.so.1.0)
                #14 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #15 0x00000000e6e526b2 method_call (libpython3.4.so.1.0)
                #16 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #17 0x00000000e6edcbdc PyEval_CallObjectWithKeywords (libpython3.4.so.1.0)
                #18 0x00000000e6f1701b t_bootstrap (libpython3.4.so.1.0)
                #19 0x00000000e6df52b9 start_thread (libpthread.so.0)
                #20 0x00000000e6d3621e __clone (libc.so.6)

                Stack trace of thread 22633:
                #0  0x00000000e703fa20 __kernel_vsyscall (linux-gate.so.1)
                #1  0x00000000e6d2ae82 __GI_ppoll (libc.so.6)
                #2  0x00000000e454fccb ppoll (libsystemd.so.0)
                #3  0x00000000e4543d15 sd_journal_wait (libsystemd.so.0)
                #4  0x00000000e45816a6 Reader_wait (_reader.cpython-34.so)
                #5  0x00000000e6e7e156 PyCFunction_Call (libpython3.4.so.1.0)
                #6  0x00000000e6ee3205 call_function (libpython3.4.so.1.0)
                #7  0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #8  0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #9  0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #10 0x00000000e6e67267 function_call (libpython3.4.so.1.0)
                #11 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #12 0x00000000e6ee339d ext_do_call (libpython3.4.so.1.0)
                #13 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #14 0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #15 0x00000000e6ee1e8b fast_function (libpython3.4.so.1.0)
                #16 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #17 0x00000000e6e67267 function_call (libpython3.4.so.1.0)
                #18 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #19 0x00000000e6e526b2 method_call (libpython3.4.so.1.0)
                #20 0x00000000e6e426b9 PyObject_Call (libpython3.4.so.1.0)
                #21 0x00000000e6edcbdc PyEval_CallObjectWithKeywords (libpython3.4.so.1.0)
                #22 0x00000000e6f1701b t_bootstrap (libpython3.4.so.1.0)
                #23 0x00000000e6df52b9 start_thread (libpthread.so.0)
                #24 0x00000000e6d3621e __clone (libc.so.6)

                Stack trace of thread 22628:
                #0  0x00000000e703fa20 __kernel_vsyscall (linux-gate.so.1)
                #1  0x00000000e6d2adab poll (libc.so.6)
                #2  0x00000000e66dbcb2 poll (select.cpython-34.so)
                #3  0x00000000e6e7e156 PyCFunction_Call (libpython3.4.so.1.0)
                #4  0x00000000e6ee3205 call_function (libpython3.4.so.1.0)
                #5  0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #6  0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #7  0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #8  0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #9  0x00000000e6ee1e8b fast_function (libpython3.4.so.1.0)
                #10 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #11 0x00000000e6ee1efa fast_function (libpython3.4.so.1.0)
                #12 0x00000000e6ee1e8b fast_function (libpython3.4.so.1.0)
                #13 0x00000000e6ee3b61 PyEval_EvalCodeEx (libpython3.4.so.1.0)
                #14 0x00000000e6ee3bf6 PyEval_EvalCode (libpython3.4.so.1.0)
                #15 0x00000000e6efd862 run_mod (libpython3.4.so.1.0)
                #16 0x00000000e6eff880 PyRun_FileExFlags (libpython3.4.so.1.0)
                #17 0x00000000e6f00ada PyRun_SimpleFileExFlags (libpython3.4.so.1.0)
                #18 0x00000000e6f015ad PyRun_AnyFileExFlags (libpython3.4.so.1.0)
                #19 0x00000000e6f14e80 run_file (libpython3.4.so.1.0)
                #20 0x0000000001caf953 main (python3.4)
                #21 0x00000000e6c5c7e6 __libc_start_main (libc.so.6)
                #22 0x0000000001caf9d8 _start (python3.4)

```

**Comment by sebres** [comment]
> 32 / 64 bits

It doesn't matter

> python3.4 sometimes segfaults

I did not understand, your default python version (under that also fail2ban runs) is 3.4 or 3.5?
```
python -V
```

**Comment by sebres** [comment]
@AndrzejL-eu I see, you are using `systemd` as backend. Can you check starting multiple jails using another backend, for example `polling`. Just to exclude this backend as point of failure.

**Comment by kakra** [comment]
systemd was there before fail2ban (I used denyhosts before). Thus, the system completely lacks traditional syslog files (this design is intentional). It would require a lot of reconfiguration to make polling useful, tho I have no doubt it would work then. I'm also pretty confident this is a problem in python's thread handling, I think there is a pull request trying to fix this by using another forking approach. Apparently, it has not been merged yet and has conflicts - and there is no branch based on stable (I'd try that otherwise).

BTW: `python -V` show 2.7.10 for me... fail2ban is using another python version:

```
$ python3.4 -V
Python 3.4.3
$ head -1 $(which fail2ban-server)
#!/usr/bin/python3.4
```

This is a result of Gentoo's multi-slot installation and PYTHON_SINGLE_TARGET. It was installed with single target set to 3.4. The default main python version is still 2.7 due to many scripts relying on that.

I think I tried polling before to rule that out, tho most rules fail simply because the logs are not there. However, it ran without the threading error (but showed other errors clearly pointing to missing log files).

**Comment by kakra** [comment]
>> 32 / 64 bits
> It doesn't matter

Well, it should matter - you may want to think about it: systemd journal reads journals by using mmap. When it is watching multiple journals (split by uid and size-based rotation) at once it may segfault because there is not enough address space available - or it simply ignores reading some of the journal files (which was my issue with journalctl: sometimes it showed all messages, sometimes it missed most). With the 64bit kernel space, this is much more relaxed since every user land process has its own complete 32bit address space wihout sharing it with the rest of the system (except the bits libraries include). Keep in mind: I use 64bit kernel with 32bit userland (for historical reasons) otherwise - in a pure 64bit world - there would be no real issues with address space. At least, this mmap issue is true for journalctl (it's documented in the issue tracker of systemd), and also for systemd-journald which writes to the logs. I suppose, since the journald python adapter probably works similar to journalctl, it undergoes the same limitations introduced by mmap. This is also not a problem with available memory (as it is mmap), the system had 12GB to use with 32bit kernel (using PAE), now it can use 64bit natively - it's a pure address space problem. This is covered by the observation that systemd-journald has no more swapping and IO latency issues now, IO has gone from megabytes to minutes down to almost zero on the journal partition (according to munin). The system no longer does intensive swapping and uses less RAM of its 12 GB (also according to munin). IO wait has gone from >50% down to <20%.

Since upgrading to 64bit kernel, I had no further issues with journalctl and systemd-journald. With 32bit kernel, I could use at most one jail. Now I can use at most 2 jails which is the same issue as reported in #969 so I suppose while my first problem may be related to mmap alone, it is now the already known threading problem of python (or maybe systemd-python).

So maybe the problem is simply how the journal's mmap'ing interacts with pythons threading implementation, and it should be reworked - as done in one of the pull requests/branches. However, I couldn't figure out which commits exactly relate to the reimplementation of forking processes. Here is the link which is supposed to implement "the fix":

https://github.com/sebres/fail2ban/tree/f2b-perfom-prepare-716-cs (see above, comment by @sebres)

**Comment by kakra** [comment]
One other idea: Maybe the threading approach is not the best regarding how systemd journal handles log access (by using mmap): All threads share the same address space, a 32bit address space will be exhausted very early. Thus, it should not be reproducible with a 64bit python, tho I'm having no opportunity currently to test this.

I'm not sure how fail2ban's threading/forking model currently works. But if it first creates new threads, then opens the journald connection, it's probably that why it fails. I'm not much into threading architecture regarding memory and address space, maybe mmaps simply cannot be shared that way after forking a new thread and that is why it fails. Maybe, a multi-process approach would work much better in case of systemd backend.

**Comment by sebres** [comment]
@kakra Well if my branch fixes this issue for you, it does this indirectly (unintentionally:))).
I am not convinced that you both have the same problem.
Because if I correct understood the original issue of @AndrzejL-eu, it describes the impossibility to start more as one jail thread (aka backend listener). I underline "starting of thread self". Of course in the constructor (or by initialize) of the jail thread, that will initialize backend, filters etc also. But all this should be done with catch the real error, occurred in the thread. Here goes something wrong, because according to log file, the thread will just not created. And we not see the real error at all.

So the question would be indeed "Whether this branch helps to solve the original issue for @AndrzejL-eu constellation?"

> So maybe the problem is simply how the journal's mmap'ing interacts with pythons threading implementation.
> Maybe the threading approach is not the best regarding how systemd journal handles log access (by using mmap)

I'm not developer of the systemd backend, but I think such multi-threaded children's diseases should be prevented by the GIL, resp. correct solution inside the systemd.journal module. On all cases it may be an issue of developers of this module (at least with regard to the documentation, how it should be proper used multi-threaded).
<hr>
Another thing, what always bothered me, is imho the "wrong" multi-threaded conception at all, that currently implemented in fail2ban. Currently if you've 50 jails - you've 50 threads, what is not a problem if all are sleeping. But it is totally not suitable for high-load systems (think about GIL, context switch, etc.)
I will try to rewrite jail pool with configurable thread pool, where you can start fail2ban for example using only `N` worker (filter listener), where all the `M` jails (with filter) can be (auto or manual) assigned to this pool. (Is meant `N` should be much smaller than `M`, ideally N=1,2 per backend).
So you could configure only 1 filter listener for the systemd backend, to process a 50 jails used this backend.
The (threaded) conception would be:
```
[1 Server*]
    |
[N Filter Listener*] <-> [M Jails] <-> [K=1 Action Thread*]

* threaded modules
```
But I'm pretty sure that this solution (branch) remains forever my fork (will never merged here)...

**Comment by yarikoptic** [comment]
FWIW, 
- I also consider current approach inefficient and hard to scale while proliferating many threads while also enjoying all the GIL problems etc
- If rewritten, I wonder if there could be only 1 listener per backend and then use subscriber model so that jails subscribe to changes for their files (some times the same file used by multiple jails) and processing then happens in more parallel fashion (multiprocessing) may be whole architecture being done in event driven fashion one way or another
- I do not oppose merging anything which could be sensibly reviewed and adjusted.  Of cause if this refactoring is yet one more on top of another hundred of commits, it would be impossible to accomplish that

**Comment by sebres** [comment]
Do not misunderstand: the above mentioned branch (f2b-perfom-prepare-716-cs) had not my new threaded conception (because not yet ready), it's still my private branch.
> I wonder if there could be only 1 listener per backend 

Since I've implemented event driven functionality (observer in #716) it is very easy to make such things with only 1 listener.

> I do not oppose merging anything which could be sensibly reviewed and adjusted

Well, although I understand that, but can not contribute this way, because we understand the word "sensibly" differently resp. I'm not ready to invest so much time in to development, and then discover, that the reviewers in turn unwilling are also time to invest (although this time would be not approximately comparable to the development time). This way, ex. to compensate with my activities the lack of time of the reviewer (to shift it on my shoulder through split the PR resp. extracts some features or cross merge etc.), is imho not acceptable (and grew to this problem not through me).

**Comment by yarikoptic** [comment]
We did our best with @kwirk to review #716 and did invest considerable time in that! Please don't state that we didn't.
FWIW it was you who suggested to split it into two (https://github.com/fail2ban/fail2ban/pull/716#issuecomment-65064050) and you found yourself hard to follow what is already addressed or not addressed (https://github.com/fail2ban/fail2ban/pull/716#issuecomment-118391768) which in turn is simply due to monumental size of the PR and mixing in various bug fixes and refactorings.  

If you would like I can try to help to split some parts out into separate PRs. e.g. what if I start by taking let's say seekToTime and furnish it as a a separate PR?  I can try to preserve your authorship over those contributions (thanks git).

**Comment by sebres** [comment]
> it was you who suggested to split it into two 

And I've tried to do that, thus for example was born a branch #1116 (and ban-time-incr rebased on to it) and then the newer version of it #1321. Because I can't make my further development on the current master (too different, too slow, too buggy, etc), the problem grew and grew, and I can't split dependent changes (all not dependent are already successfully merged into master and I had cross-merge every now and then).

In the mean time, my ban-time-incr branch totally rebased over #1321 and your proposal is at least one year "out of date" (but nevertheless thanks)...

Now we should merge (and I insist on exact such a sequence):
- sebres/f2b-perfom-prepare-716 (prev. #1116);
- sebres/f2b-perfom-prepare-716-cs (prev. #1321) - rewritten client/server functionality, ca. 8 commits over sebres/f2b-perfom-prepare-716;
- sebres/ban-time-incr (prev. #716) - merge with ca 10 commits over sebres/f2b-perfom-prepare-716-cs;

And then I should cross-merge / rebase ca. 5 (more or less large) my own private branches with further development (because as I know you, all this above will be not merged without changes =)).

**Comment by yarikoptic** [comment]
```
$> git diff origin/master...github_sebres_fail2ban/f2b-perfom-prepare-716 --stat | tail -n 1 
 48 files changed, 2258 insertions(+), 705 deletions(-)
```
heh... as we discussed in #1116 I wish it was multiple PRs, but ok -- let's make another attempt -- send it as a PR? meanwhile right away consider my question which I have raised before -

- you introduce getTime2 (in addition to previous getTime) -- it is pretty much a decorator for the old getTime allowing also for providing previously matched results.  Why not just to fold that logic at the beginning of getTime?  getTime2 name is not descriptive, blows up API for no reason.  While at it, stay closer to PEP8 (`timeMatch = None` -> `timeMatch=None`) in function signature
- please go through diff and add docs to new functions... e.g. seekToTime -- "accuracy" -- what its units and role?
- in seekToTime -- seeking is done on bytes level ,not lines level,  so by virtue of that "line = container.readline()" might return some truncated line which could match the date but incorrectly (e.g. if timestamps starts with sec:min:hh: and it is 23:45:01  but readline would return 3:45:01)... imho better read one line and discard it and then go to the next one which you would know to be complete for sure
(see -- that is why I wanted this as a separate PR so we could iron out issues for 1 particular contribution at a time)

and yes, it will take me a while to get through it and yes it would take me quite a bit of comments/answer to figure out and yes we might get lost again because again of its size.

**Comment by sebres** [comment]
BTW seekToTime always seeks to begin of line

**Comment by yarikoptic** [comment]
On Mon, 07 Mar 2016, Serg G. Brester wrote:

>    BTW seekToTime always seeks to begin of line

aha... indeed but not seekToTime but seek does that ;-P  cool, that one
"resolved" ;)

**Comment by sebres** [comment]
closed as duplicate of #969

Note please https://github.com/fail2ban/fail2ban/issues/969#issuecomment-240350774
